apollo-api/src/iosMain/kotlin/com/apollographql/apollo/api/ThreadSafeMap.kt

```diff
@@ -0,0 +1,36 @@
+package com.apollographql.apollo.api
+
+import kotlinx.cinterop.StableRef
+import platform.darwin.DISPATCH_QUEUE_SERIAL
+import platform.darwin.dispatch_queue_create
+import platform.darwin.dispatch_queue_t
+import platform.darwin.dispatch_sync
+import kotlin.native.concurrent.AtomicReference
+import kotlin.native.concurrent.freeze
+
+actual class ThreadSafeMap<K, V> {
+  private val stableRef = StableRef.create(mutableMapOf<K, V>())
+  private val queue = dispatch_queue_create(label = "map", attr = DISPATCH_QUEUE_SERIAL as dispatch_queue_t)
+
+  actual fun getOrPut(key: K, defaultValue: () -> V): V {
+    defaultValue.freeze()
+    val reference = AtomicReference<V?>(null)
+    dispatch_sync(
+        queue = queue,
+    ) {
+      initRuntimeIfNeeded()
+
+      reference.value = stableRef.get().getOrPut(key, defaultValue).freeze()
+    }
+
+    val result = reference.value
+    reference.value = null
+    return result!!
+  }
```

Comment:
sorry but I'm not sure I'm following what is going in here.
Map was created outside the queue, we de-refrence it in the queue, block caller thread of `getOrPut` and mutate it outside of queue.

Is intention to make `isolate` pattern here? When accessing map confined to specific thread?

Reply:
I'm not 100% sure either. I wanted to do push this before renaming everything to apollo3 but this'll have to be properly tested and most likely reworked.

Reply:
It's working now with https://github.com/apollographql/apollo-android/pull/2924/commits/39d3ccd860f9b987c687510b36ae89d29ce211b9. The previous code was indeed completely broken 🙃 .